### PR TITLE
HIPBlasLt integration

### PR DESCRIPTION
**Pull request by zstreet87**
Requires `USE_HIPBLASLT=1`

`HIPBLASLT_LOG_LEVEL=5 python test_linalg.py -v TestLinalgCUDA.test_corner_cases_of_cublasltmatmul_cuda_float32`

```bash
Test results will be stored in test-reports/python-unittest/test_linalg

Running tests...
----------------------------------------------------------------------
  test_corner_cases_of_cublasltmatmul_cuda_float32 (__main__.TestLinalgCUDA) ... [2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_create] matmulDesc[out]=0x7fffa3161bd8 computeType=COMPUTE_32F scaleType=R_32F
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x80b5880 attr=MATMUL_DESC_TRANSA buf=0x7fffa3161bc0 sizeInBytes=4 bufData=0x70
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x80b5880 attr=MATMUL_DESC_TRANSB buf=0x7fffa3161bc4 sizeInBytes=4 bufData=0x6f
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x80b5880 attr=MATMUL_DESC_EPILOGUE buf=0x7fffa3161bc8 sizeInBytes=4 bufData=0x4
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x80b5880 attr=MATMUL_DESC_BIAS_POINTER buf=0x7fffa3161ba0 sizeInBytes=8 bufData=0x9ae00000
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161bd8 type=R_32F rows=2400 cols=128 ld=2400
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161bd8 type=R_32F rows=2400 cols=2048 ld=2400
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161bd8 type=R_32F rows=128 cols=2048 ld=128
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_create] matmulPref[out]=0x80b4dd0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_set_attribute] matmulPref=0x80b4dd0 attr=1 buf=0x7fffa3161bd0 sizeInBytes=8 data=1048576
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_algo_get_heuristic] returnAlogCount=1
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul] A=0x7efa9aac0000 Adesc=0x743a5f0 B=0x7efa99800000 Bdesc=0x80ef6b0 C=0x7efa9ae00200 Cdesc=0x80ef8f0 D=0x7efa9ae00200 Ddesc=0x80ef8f0 computeDesc=0x80b5880 algo=0x7fffa3161c30 workSpace=0x7efa99400000 workSpaceSizeInBytes=1048576 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Trace][rocblaslt_matmul] A=0x7efa9aac0000 Adesc=[type=R_32F rows=2400 cols=128 ld=2400] B=0x7efa99800000 Bdesc=[type=R_32F rows=2400 cols=2048 ld=2400] C=0x7efa9ae00200 Cdesc=[type=R_32F rows=128 cols=2048 ld=128] D=0x7efa9ae00200 Ddesc=[type=R_32F rows=128 cols=2048 ld=128]
computeDesc=[computeType=COMPUTE_32F scaleType=R_32F transA=OP_T transB=OP_N epilogue=EPILOGUE_BIAS biasPointer=0x9ae00000] workSpace=0x7efa99400000 workSpaceSizeInBytes=1048576 alpha=1 beta=0 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_destroy] matmulPref=0x80b4dd0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x80ef8f0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x80ef6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x743a5f0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_destroy] matmulDesc=0x80b5880
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_create] matmulDesc[out]=0x7fffa3161d18 computeType=COMPUTE_32F scaleType=R_32F
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x743a5f0 attr=MATMUL_DESC_TRANSA buf=0x7fffa3161d00 sizeInBytes=4 bufData=0x70
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x743a5f0 attr=MATMUL_DESC_TRANSB buf=0x7fffa3161d04 sizeInBytes=4 bufData=0x6f
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x743a5f0 attr=MATMUL_DESC_EPILOGUE buf=0x7fffa3161d08 sizeInBytes=4 bufData=0x4
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x743a5f0 attr=MATMUL_DESC_BIAS_POINTER buf=0x7fffa3161ce0 sizeInBytes=8 bufData=0x9ae00200
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161d18 type=R_32F rows=2400 cols=128 ld=2400
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161d18 type=R_32F rows=2400 cols=2048 ld=25272
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161d18 type=R_32F rows=128 cols=2048 ld=128
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_create] matmulPref[out]=0x897d760
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_set_attribute] matmulPref=0x897d760 attr=1 buf=0x7fffa3161d10 sizeInBytes=8 data=1048576
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_algo_get_heuristic] returnAlogCount=1
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul] A=0x7efa94800000 Adesc=0x80b5880 B=0x7efa7b8156d0 Bdesc=0x80b5a60 C=0x7efa9ae00400 Cdesc=0x77d9cc0 D=0x7efa9ae00400 Ddesc=0x77d9cc0 computeDesc=0x743a5f0 algo=0x7fffa3161d70 workSpace=0x7efa99400000 workSpaceSizeInBytes=1048576 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Trace][rocblaslt_matmul] A=0x7efa94800000 Adesc=[type=R_32F rows=2400 cols=128 ld=2400] B=0x7efa7b8156d0 Bdesc=[type=R_32F rows=2400 cols=2048 ld=25272] C=0x7efa9ae00400 Cdesc=[type=R_32F rows=128 cols=2048 ld=128] D=0x7efa9ae00400 Ddesc=[type=R_32F rows=128 cols=2048 ld=128] computeDesc=[computeType=COMPUTE_32F scaleType=R_32F transA=OP_T transB=OP_N epilogue=EPILOGUE_BIAS biasPointer=0x9ae00200] workSpace=0x7efa99400000 workSpaceSizeInBytes=1048576 alpha=1 beta=0 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_destroy] matmulPref=0x897d760
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x77d9cc0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x80b5a60
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x80b5880
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_destroy] matmulDesc=0x743a5f0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_create] matmulDesc[out]=0x7fffa3161d18 computeType=COMPUTE_32F scaleType=R_32F
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x7d5edd0 attr=MATMUL_DESC_TRANSA buf=0x7fffa3161d00 sizeInBytes=4 bufData=0x70
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x7d5edd0 attr=MATMUL_DESC_TRANSB buf=0x7fffa3161d04 sizeInBytes=4 bufData=0x6f
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x7d5edd0 attr=MATMUL_DESC_EPILOGUE buf=0x7fffa3161d08 sizeInBytes=4 bufData=0x4
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x7d5edd0 attr=MATMUL_DESC_BIAS_POINTER buf=0x7fffa3161ce0 sizeInBytes=8 bufData=0x9ae00000
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161d18 type=R_32F rows=2400 cols=128 ld=25272
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161d18 type=R_32F rows=2400 cols=2048 ld=2400
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161d18 type=R_32F rows=128 cols=2048 ld=128
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_create] matmulPref[out]=0x80ef6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_set_attribute] matmulPref=0x80ef6b0 attr=1 buf=0x7fffa3161d10 sizeInBytes=8 data=1048576
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_algo_get_heuristic] returnAlogCount=1
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul] A=0x7efa949416d0 Adesc=0x7d5eea0 B=0x7efa99800000 Bdesc=0x7d5eed0 C=0x7efa9ae00200 Cdesc=0x7d60490 D=0x7efa9ae00200 Ddesc=0x7d60490 computeDesc=0x7d5edd0 algo=0x7fffa3161d70 workSpace=0x7efa99400000 workSpaceSizeInBytes=1048576 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Trace][rocblaslt_matmul] A=0x7efa949416d0 Adesc=[type=R_32F rows=2400 cols=128 ld=25272] B=0x7efa99800000 Bdesc=[type=R_32F rows=2400 cols=2048 ld=2400] C=0x7efa9ae00200 Cdesc=[type=R_32F rows=128 cols=2048 ld=128] D=0x7efa9ae00200 Ddesc=[type=R_32F rows=128 cols=2048 ld=128] computeDesc=[computeType=COMPUTE_32F scaleType=R_32F transA=OP_T transB=OP_N epilogue=EPILOGUE_BIAS biasPointer=0x9ae00000] workSpace=0x7efa99400000 workSpaceSizeInBytes=1048576 alpha=1 beta=0 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_destroy] matmulPref=0x80ef6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x7d60490
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x7d5eed0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x7d5eea0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_destroy] matmulDesc=0x7d5edd0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_create] matmulDesc[out]=0x7fffa3161bd8 computeType=COMPUTE_32F scaleType=R_32F
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x897d760 attr=MATMUL_DESC_TRANSA buf=0x7fffa3161bc0 sizeInBytes=4 bufData=0x70
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x897d760 attr=MATMUL_DESC_TRANSB buf=0x7fffa3161bc4 sizeInBytes=4 bufData=0x6f
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x897d760 attr=MATMUL_DESC_EPILOGUE buf=0x7fffa3161bc8 sizeInBytes=4 bufData=0x4
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_set_attribute] matmulDesc=0x897d760 attr=MATMUL_DESC_BIAS_POINTER buf=0x7fffa3161ba0 sizeInBytes=8 bufData=0x9ae00200
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161bd8 type=R_32F rows=131071 cols=16 ld=131071
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161bd8 type=R_32F rows=131071 cols=32 ld=131071
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_create] matLayout[out]=0x7fffa3161bd8 type=R_32F rows=16 cols=32 ld=16
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_create] matmulPref[out]=0x81442f0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_set_attribute] matmulPref=0x81442f0 attr=1 buf=0x7fffa3161bd0 sizeInBytes=8 data=1048576
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_algo_get_heuristic] returnAlogCount=1
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul] A=0x7efa94800000 Adesc=0x80b5880 B=0x7efa7b800000 Bdesc=0x7d5f8c0 C=0x7efa9ae00400 Cdesc=0x743a5f0 D=0x7efa9ae00400 Ddesc=0x743a5f0 computeDesc=0x897d760 algo=0x7fffa3161c30 workSpace=0x7efa9ae00c00 workSpaceSizeInBytes=1048576 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Trace][rocblaslt_matmul] A=0x7efa94800000 Adesc=[type=R_32F rows=131071 cols=16 ld=131071] B=0x7efa7b800000 Bdesc=[type=R_32F rows=131071 cols=32 ld=131071] C=0x7efa9ae00400 Cdesc=[type=R_32F rows=16 cols=32 ld=16] D=0x7efa9ae00400 Ddesc=[type=R_32F rows=16 cols=32 ld=16] computeDesc=[computeType=COMPUTE_32F scaleType=R_32F transA=OP_T transB=OP_N epilogue=EPILOGUE_BIAS biasPointer=0x9ae00200] workSpace=0x7efa9ae00c00 workSpaceSizeInBytes=1048576 alpha=1 beta=0 stream=0x73ed6b0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_preference_destroy] matmulPref=0x81442f0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x743a5f0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x7d5f8c0
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matrix_layout_destory] matLayout=0x80b5880
[2023-06-12 21:16:46][HIPBLASLT][690719][Api][rocblaslt_matmul_desc_destroy] matmulDesc=0x897d760
ok (0.914s)

----------------------------------------------------------------------
Ran 1 test in 0.917s

OK

Generating XML reports...
Generated XML report: test-reports/python-unittest/test_linalg/TEST-TestLinalgCUDA-20230612211645.xml

```

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -2500,11 +2500,11 @@ def test_invariance_error_spectral_decompositions(self, device, dtype):
             Q = torch.linalg.eigh(A).eigenvectors
             Q.sum().abs().backward()
 
-<<<<<<< HEAD
+# <<<<<<< HEAD
     @skipCUDAIfNoCusolverAndNoHipsolver  # MAGMA backend doesn't work in this case
-=======
-    @skipCUDAIfNoCusolver  # MAGMA backend doesn't work in this case
->>>>>>> downstream_main
+# =======
```
Comment (zstreet87) [inline test/test_linalg.py:2505]: This should be addressed.

File: test/test_linalg.py [hunk 2]
```diff
@@ -2500,11 +2500,11 @@ def test_invariance_error_spectral_decompositions(self, device, dtype):
...
+#     @skipCUDAIfNoCusolver  # MAGMA backend doesn't work in this case
+# >>>>>>> downstream_main
```
Comment (zstreet87) [inline test/test_linalg.py:2507]: This should be addressed.

**Comment by pruthvistony** [comment]
@zstreet87 , 
Can you rebase this PR. Since hipify correction related to API_SOLVER changes have been merged.

**Comment by zstreet87** [comment]
> @zstreet87 , Can you rebase this PR. Since hipify correction related to API_SOLVER changes have been merged.

This PR has been merged into the 5.7_internal_testing branch?

**Comment by jeffdaily** [comment]
How do we retrigger CI?  Something went wrong.

**Comment by zstreet87** [comment]
@jithunnair-amd ?

**Review by jithunnair-amd — APPROVED** [review]

**Comment by jithunnair-amd** [comment]
@zstreet87 The PyTorch build failed: http://rocmhead:8080/job/pytorch/job/pytorch-build/820/console
```
11:01:43 Jun 16 16:01:43 CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
11:01:43 Jun 16 16:01:43 Please set them or make sure they are set and tested correctly in the CMake files:
11:01:43 Jun 16 16:01:43 PYTORCH_HIPBLASLT_LIBRARIES
11:01:43 Jun 16 16:01:43     linked by target "torch_hip" in directory /var/lib/jenkins/pytorch/caffe2
```

**Comment by zstreet87** [comment]
@jeffdaily thoughts? I'm pretty sure the error @jithunnair-amd is reporting is because hipblaslt isn't installed but 5.5 should have it...? I'll double check

**Comment by zstreet87** [comment]
@jithunnair-amd I'm not sure what to make of the ci failure...

**Comment by jithunnair-amd** [comment]
The pytorch build succeeded: http://rocmhead:8080/job/pytorch/job/pytorch-build/826/console.
The pytorch tests are failing due to an unrelated issue. Merging this PR.

**Comment by zstreet87** [comment]
Thanks @jithunnair-amd!

I'll make a PR for upstream now but now we can move our internal JIRA tickets along.
cc @jeffdaily @sunway513

**Pull request merged**